staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -379,6 +379,10 @@ type ListOptions struct {
 	// This field is not supported when watch is true. Clients may start a watch from the last
 	// resourceVersion value returned by the server and not miss any modifications.
 	Continue string `json:"continue,omitempty" protobuf:"bytes,8,opt,name=continue"`
+	// A subset of fields to return
+	FieldMask []string `json:"fieldMask,omitempty" protobuf:"bytes,9,rep,name=fieldMask"`
+	// A subset of fields which are considered relevant for watch notifications
+	WatchMask []string `json:"watchMask,omitempty" protobuf:"bytes,10,rep,name=watchMask"`
```

Comment:
I think 99.5% of this problem is justifying the API. The implementation should be pretty straightforward, we already have code in all the places necessary.

Eventually we need to be able to do this filtering in the storage system, so whatever we come up with needs to be relatable to the data type without having the schema readily available, or we introduce the requirement that the storage layer needs to understand the schema.

For the specific thing here, I don't think watchMask and fieldMask should both exist, certainly not in the same format. I don't think a []string is the right type for identifying fields.

I don't think this API can be developed independently from the query API. What you have here is the SELECT clause of a SQL query, but people will also want to write the WHERE clause.

Reply:
> I think 99.5% of this problem is justifying the API.

And that's a non-trivial sell.  How does this contrast with different `Accept` headers like @smarterclayton started with server-side printing.  We had long discussed using it for objectmeta only

Reply:
I've only implemented metadata fields, but I do think there are probably a handful of non-metadata fields we would want to support.  e.g. I think some CNI providers want to watch Node spec.podCIDR.  I think it'll be a lively discussion as to whether we would ever want to support _arbitrary_ field masks.

But yes, we could see this as an extension of the same concept.  If you think it's better to express this in the Accept header, that is fine with me :-)

This would be something more like: 
 `Accept: application/json;g=meta.k8s.io;v=v1beta1;fieldMask=spec.podCidr,metadata.name`